src/client/signatures/canonical_json/src/ser.rs

```diff
@@ -0,0 +1,595 @@
+use serde::ser::{Serialize};
+use serde_json::ser::{Formatter, CharEscape};
+use std::io::Write;
+use log::{debug, error};
+
+struct JSONFormatter {}
+
+impl Formatter for JSONFormatter {
+    fn write_f64<W: ?Sized>(&mut self, writer: &mut W, value: f64) -> Result<(), std::io::Error> where
+        W: Write,
+    {
+        format_number(writer, value, 10.0_f64.powf(21.0), 10.0_f64.powf(-6.0))?;
+        Ok(())
+    }
+
+    fn write_char_escape<W: ?Sized>(&mut self, writer: &mut W, char_escape: CharEscape) -> Result<(), std::io::Error> where
+        W: Write, 
+    {
+        match char_escape {
+            CharEscape::Quote => {
+                writer.write("\"".as_bytes())?; 
+            },
+            CharEscape::ReverseSolidus => {
+                writer.write("\\\\".as_bytes())?;
+            },
+            CharEscape::LineFeed => {
+                writer.write("\n".as_bytes())?;
+            },
+            CharEscape::Tab => {
+                writer.write("\t".as_bytes())?;
+            },
+            CharEscape::CarriageReturn => {
+                writer.write("\r".as_bytes())?;
+            },
+            CharEscape::Backspace => {},
+            CharEscape::FormFeed => {},
+            CharEscape::Solidus => {},
+            CharEscape::AsciiControl (_number) => {}
+        }
+
+        Ok(())
+    }
+
+    fn write_string_fragment<W: ?Sized>(&mut self, writer: &mut W, fragment: &str) -> Result<(), std::io::Error> where
+        W: Write, 
+    {
+        let formatted_string = format!("{}", fragment).escape_default().to_string().replace("\\\"", "\"");
+        format_unicode_in_string(writer, formatted_string)?;
+        Ok(())
+    }
+}
+
+fn format_number<W: ?Sized>(writer: &mut W, number: f64, lower_bound: f64, upper_bound: f64) -> Result<(), std::io::Error> where W: Write, {
+
+    if (0.0 < number && number < upper_bound) || (number >= lower_bound) {
+        debug!("converting number {} to scientific notation", number);
+        let number_string = format!("{:e}", number);
+        debug!("formatted_number {}", number_string);
+        let mut prev_char = '\0';
+        for curr_char in number_string.chars() {
+            if prev_char == 'e' && curr_char != '-' {
+                // its a positive exponent
+                writer.write("+".as_bytes())?;
+            }
+
+            writer.write(curr_char.to_string().as_bytes())?;
+            prev_char = curr_char;
+        }
+
+        return Ok(());
+    }
+
+    debug!("returning number {} without scientifc notation", number);
+    writer.write(&format!("{}", number).into_bytes())?;
+    Ok(())
+}
+
+/// looking for \u{X} \u{XX}, \u{XXX}, \u{XXXX} to remove the curly braces
+fn format_unicode_in_string<W: ?Sized>(writer: &mut W, serialized_string: String) -> Result<(), std::io::Error> where W: Write, {
+
+    let mut string_iter = serialized_string.chars().peekable();
+
+    while let Some(curr_char) = string_iter.next() {
```

Comment:
nit: Tighten the vertical whitespace hre, no blank lanes after loop/block starts, please.

Reply:
Better: run `rustfmt`.

Reply:
done